### PR TITLE
Color Picker: Creating a custom color picker

### PR DESCRIPTION
**Pull request by youknowriad**
We definitely need to find a better way to style the "Random color picker" but this PR tries to improve the color-picker with a clear color option and a random color picker.

### CODE REVIEW

**Comment by jasmussen** [comment]
I think this seems good. Thank you for working on this.

Seeing some style issues:

screen shot 2017-08-14 at 16 31 36

And a fun outline:

screen shot 2017-08-14 at 16 30 44

I like the "no swatch" option, btw ;)

**Comment by youknowriad** [comment]
Should we merge this?

**Comment by jasmussen** [comment]
I like it, I think it's good!

Matías wasn't as much a fan of the red circle with line through it as I was (I'm the one that suggested it). He suggested a "Remove color" link.

So instead of:

screen shot 2017-08-17 at 10 44 01

maybe something like this?

screen-shot-2017-08-17-at-10 44 01

Maybe "Remove" is enough?

But either way I have no strong opinion, I think it's good to get this in soon though.

**Comment by mtias** [comment]
@jasmussen any clue why the first set of colors has the rainbow-circle slightly cut off at the top but not the bottom one?

**Comment by jasmussen** [comment]
> any clue why the first set of colors has the rainbow-circle slightly cut off at the top but not the bottom one?

Huh, that is a doozy. I _sense_ that it must be some subpixel stuff, but I can't figure it out. Check this out, though:

huh

If you apply a 20px top margin on the first container of blogs, it's uncropped. If that margin is 0-19px, it's cropped.

**Comment by mtias** [comment]
@afercia added a few aria-roles, but I wonder what you think of `aria-current` and its usage. Would it work if the group is not a list? Is there a better approach for this one? Should we also add a label to the main palette wrapper stating which color is currently selected?

**Comment by youknowriad** [comment]
@mtias I added an `eslint-disable` comment. I think we should merge and reconsider once @afercia gets back from AFK :)

**Comment by mtias** [comment]
👍

**Review by mtias — APPROVED** [review]

File: blocks/color-palette/index.js [hunk 1]
```diff
@@ -1,26 +1,94 @@
...
+					const className = classnames( 'blocks-color-palette__item', { 'is-active': value === color } );
+					/* Disable reason: aria-current seems like a valid aria-prop and well-suited for this */
+					/* eslint-disable jsx-a11y/aria-props */
+					return (
+						<div key={ color } className="blocks-color-palette__item-wrapper">
+							<button
+								className={ className }
+								style={ style }
+								onClick={ () => onChange( value === color ? undefined : color ) }
+								aria-label={ __( 'Color: ' ) + color }
```
Comment (aduth) [inline blocks/color-palette/index.js]: We should use variable substitution here:

```js
sprintf( __( 'Color: %s', color )
```

>- Do not leave leading or trailing whitespace in a translatable phrase.
>- Add the variables as placeholders to the string as in some languages the placeholders change position.

https://developer.wordpress.org/themes/functionality/internationalization/#best-practices-for-writing-strings

**Comment by afercia** [comment]
> @afercia added a few aria-roles, but I wonder what you think of aria-current and its usage. Would it work if the group is not a list? Is there a better approach for this one? Should we also add a label to the main palette wrapper stating which color is currently selected?

Bit late, but it's on my list 🙂 @mtias I think `aria-current` is probably not the best option here, but I haven't checked in depth. Will try to test the new color picker soon.

**Comment by mtias** [comment]
Yes, I read contradicting statements around aria-current. :)

**Comment by afercia** [comment]
> Yes, I read contradicting statements around aria-current. :)

Yep, always better read the documentation before relying on other sources... 🙂 

https://www.w3.org/TR/wai-aria-1.1/#aria-current
> Authors should not use the aria-current attribute as a substitute for aria-selected etc. etc.

Also, aria-current can have only the values specified in the recommendation:
page, step, location, date, time, true, and false
Instead, currently it's `aria-current="Selected color"` which is invalid...

However, I'd say in this case the best option would probably be `aria-pressed`
https://www.w3.org/TR/wai-aria-1.1/#aria-pressed
> Indicates the current "pressed" state of toggle buttons. Will open a new issue, if there isn't one already.

These controls are actually buttons and only one button at a time can be selected within its set, so they're "toggle buttons".

**Comment by mtias** [comment]
@afercia the most important to me is how—and where—to announce which color is currently selected.

**Comment by afercia** [comment]
@mtias that's obvious 🙂 

https://www.w3.org/TR/wai-aria-1.1/#aria-current
>  Any value not included in the list of allowed values should be treated by assistive technologies as if the value true had been provided.

What screen readers actually announce is "current item", as if the value provided was "true".

screen shot 2017-09-05 at 19 14 47

To recap:

`Color {brief pause} number eight ed one f c, current item, button`

where
- the first part comes from the aria-label value `Color: #8ed1fc`
- "current item" comes from aria-current as if it was "true", the actual value "Selected color" gets totally ignored
- "button" is the element type screen readers announce by default

**Comment by mtias** [comment]
So `Color: value, current item` would make sense then? How would you know, without getting into the individual buttons, which color is selected?

**Comment by afercia** [comment]
I think our communication is failing a bit 🙁

**Pull request merged**